enhancements/security/scc-csi-volumes.md

```diff
@@ -0,0 +1,234 @@
+---
+title: scc-csi-volumes
+authors:
+  - "@adambkaplan"
+reviewers:
+  - "@bparees"
+  - "@jsafrane"
+approvers:
+  - "@sttts"
+  - "@deads2k"
+creation-date: 2021-04-27
+last-updated: 2021-04-27
+status: provisional
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Security Context Constraints for CSI Volumes
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Ephemeral CSI volumes are a newer type of volume in Kubernetes, which allows content to mounted into pods with underlying ephemeral file systems.
+OpenShift Security Context Constraints (SCCs) take an all-or-nothing approach when it comes to these volumes.
+A service account/user can either mount a `csi` volume, or it cannot.
+This proposal will enhance SCCs to control which CSI drivers are allowed.
+
+## Motivation
+
+Inline CSI volumes are provided via CSI drivers that can be added to OpenShift.
+When a pod mounts an inline `csi` volume, it must specify the driver that will provide the underlying content.
+Cluster admins will need a means to identify which drivers are considered safe for use, including in restricted namespaces.
+
+Ephemeral CSI drivers are expected to be used in the following use cases:
+
+1. Mounting shared simple content access certificates used to install RHEL content.
+2. Accessing [sealed secrets](https://secrets-store-csi-driver.sigs.k8s.io/) stored in Vault, Azure Key Vault, or Google Secret Manager.
+
+### Goals
+
+* Allow cluster admins to control which ephemeral CSI drivers can be used to mount `csi` volumes.
+
+### Non-Goals
+
+* Allow `csi` volumes to be mounted in build pods.
+* Enable SCCs to limit [generic ephemeral volumes](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes)
+
+## Proposal
+
+### User Stories
+
+As an OpenShift cluster admin
+I want to control which CSI drivers can be used to mount csi volumes
+So that my workloads only use trusted CSI drivers to access content
+
+As an application developer
+I want to use the secret store CSI driver to mount csi volumes
+So that I can access sealed secrets in my workloads
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Safe for Consumption Annotation
+
+CSI drivers that provide inline CSI volumes must mark themselves as safe for consumption.
+To mark a CSI driver safe, admins or an operator will need to add the `security.openshift.io/allow-csi-volumes-restricted` annotation to the driver's `CSIDriver` object:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: CSIDriver
+annotations:
+  security.openshift.io/allow-csi-volumes-restricted: "true"
```

Comment:
this is a pretty confusing name.  Is it allowing something? is it restricting it?  Is it allowing use of otherwise restricted things?  Restricting use of otherwise allowed things?